File: cpp/hash_table.cpp
#include "hash_table.h"
#include <filesystem>
#include <fstream>
#include <mutex>
#include <sstream>
#include <stdexcept>
#include <iostream>
#include <nlohmann/json.hpp>


using json = nlohmann::json;


PersistentHashTable::PersistentHashTable(const std::string& table_dir, bool log, int bucket_mag)
    : table_dir(table_dir), bucket_mag(bucket_mag) {
    std::filesystem::create_directories(this->table_dir);
    if (log) {
        this->log_file = this->table_dir / "log.txt";
        if (!std::filesystem::exists(this->log_file)) {
            std::ofstream(this->log_file);
        }
    }
}

PersistentHashTable::PersistentHashTable(const std::string& log_path, const std::string& table_dir, int bucket_mag)
    : table_dir(table_dir), bucket_mag(bucket_mag) {
    if (!std::filesystem::exists(log_path)) {
        throw std::runtime_error("Log path not found!");
    }

    // wipe pre-existing table dir
    std::filesystem::path tableDirPath(table_dir);
    if (std::filesystem::exists(tableDirPath)) {
        std::filesystem::remove_all(tableDirPath);
    }

    // initialize
    std::filesystem::create_directories(this->table_dir);
    this->log_file = this->table_dir / "log.txt";
    if (!std::filesystem::exists(this->log_file)) {
        std::ofstream(this->log_file);
    }

    std::ifstream logFile(log_path);
    std::string line;
    while (getline(logFile, line)) {
        std::istringstream iss(line);
        int key, value;
        char delim;
        if (!(iss >> key >> delim >> value) || delim != ',') {
            throw std::runtime_error("Invalid log format");
        }
        this->setItem(key, value);
    }
}

std::string PersistentHashTable::hash(int key) {
    std::stringstream ss;
    ss << std::hex << (key % (1 << (bucket_mag * 4)));
    return ss.str();
}

std::mutex& PersistentHashTable::getBucketLock(const std::filesystem::path& bucket_path) {
    std::lock_guard<std::mutex> lock(bucket_lock_lock);
    return bucket_lock[bucket_path.string()];
}

std::filesystem::path PersistentHashTable::getBucketPath(int key) {
    auto hashedKey = hash(key);
    auto bucketPath = table_dir / (hashedKey + ".json");
    if (!std::filesystem::exists(bucketPath)) {
        std::ofstream(bucketPath) << "{}"; // empty JSON file if not exists
    }
    return bucketPath;
}

void PersistentHashTable::log(int key, int value) {
    if (!log_file.empty()) {
        std::lock_guard<std::mutex> lock(log_lock);
        std::ofstream logStream(log_file, std::ios::app);
        logStream << key << "," << value << std::endl;
    }
}

std::filesystem::path PersistentHashTable::getLogFile() {
    return log_file;
}

int PersistentHashTable::getItem(int key) {
    auto bucketPath = getBucketPath(key);
    std::lock_guard<std::mutex> lock(getBucketLock(bucketPath));
    std::ifstream bucketFile(bucketPath);
    json bucket;
    bucketFile >> bucket;
    std::string keyStr = std::to_string(key);
    if (bucket.contains(keyStr)) {
        return bucket[keyStr];
    }
    throw std::runtime_error("Key not found");
}

void PersistentHashTable::setItem(int key, int value) {
    auto bucketPath = getBucketPath(key);
    std::lock_guard<std::mutex> lock(getBucketLock(bucketPath));
    std::ifstream bucketFileRead(bucketPath);
    json bucket;
    bucketFileRead >> bucket;
    bucketFileRead.close();
    bucket[std::to_string(key)] = value;
    std::ofstream bucketFileWrite(bucketPath);
    bucketFileWrite << bucket.dump();
    log(key, value);
}

File: python/hash_table.py
import shutil
import hashlib
import json
import threading
from pathlib import Path
from typing import Union


class PersistentHashTable:

    def __init__(self, table_dir: Union[str, Path], log: bool = True, bucket_mag: int = 3):
        """
        Initialize the PersistentHashTable object

        Set up the table directory if the table directory does not exist, or
        will reference a pre-existing table directory as-is.

        Params:
            table_dir: Path-like to directory for the hash table
            log: Flag for loggin set actions (useful for recovery)
            bucket_mag: Magnitude of number of buckets, calculation of 16^bucket_mag.
                Increasing the number of buckets can be useful for RAM constrained envs.
                Default value of 3 yields 4096 buckets.
        """
        self.table_dir = Path(table_dir)
        self.table_dir.mkdir(mode=0o777, parents=True, exist_ok=True)
        self.bucket_mag = 3
        self.bucket_lock = {}
        self.bucket_lock_lock = threading.Lock()
        self.log_lock = threading.Lock()

        if log:
            self.log_file = self.table_dir / 'log.txt'
            if not self.log_file.exists():
                self.log_file.touch(mode=0o777)
        else:
            self.log_file = None

    @classmethod
    def from_log(cls, log_path: Union[str, Path], table_dir: Union[str, Path], bucket_mag: int = 3):
        """
        Initialize the PersistentHashTable object from a log-file

        Sets up the table directory by 'replaying' the actions recorded in the log-file.
        This is useful for reloading a corrupted file system from a valid log-file.

        Params:
            log_path: Path-like to log-file for the hash table
            table_dir: Path-like to directory for the hash table
            log: Flag for loggin set actions (useful for recovery)
            bucket_mag: Magnitude of number of buckets, calculation of 16^bucket_mag.
                Increasing the number of buckets can be useful for RAM constrained envs.
                Default value of 3 yields 4096 buckets.
        """
        log_path = Path(log_path)
        if not log_path.exists():
            raise ValueError('Log path not found!')
        log_contents = log_path.read_text()

        # wipe table dir if it exists
        table_dir = Path(table_dir)
        if table_dir.exists():
            shutil.rmtree(table_dir)
        table_dir.mkdir(mode=0o775, parents=True)

        # replay log
        ht = cls(table_dir, True, bucket_mag)
        for line in log_contents.strip().split():
            k, v = line.split(',')
            ht[int(k)] = int(v)

        return ht
                
    def _hash(self, key: int) -> str:
        """
        Deterministic hashing for bucketing

        Default of 3 hexadecimal values for 4096 (2^12) buckets
        """
        return hashlib.sha256(str(key).encode()).hexdigest()[:self.bucket_mag]

    def _get_bucket_lock(self, bucket_path: Path) -> threading.Lock:
        """
        Get the lock for any given bucket
        """
        with self.bucket_lock_lock:
            if bucket_path not in self.bucket_lock:
                self.bucket_lock[bucket_path] = threading.Lock()
            return self.bucket_lock[bucket_path]

    def _get_bucket_path(self, key: int) -> Path:
        """
        Get bucket path from hashing of key

        Makes bucket path if it doesn't exist
        """
        bucket_path = self.table_dir / f'{self._hash(key)}.json'
        if not bucket_path.exists():
            json.dump({}, bucket_path.open('w'))
        return bucket_path

    def _log(self, key: int, val: int):
        """
        Log the setting of key and value
        """
        with self.log_file.open('a') as f:
            f.write(f'{key},{val}\n')

    def __getitem__(self, key: int) -> int:
        """
        Get hash table value by key
        """
        bucket_path = self._get_bucket_path(key)
        bucket_lock = self._get_bucket_lock(bucket_path)
        with bucket_lock:
            bucket = json.load(bucket_path.open('r'))
            try:
                return bucket[f'{key}']
            except Exception as e:
                raise e

    def __setitem__(self, key: int, val: int) -> None:
        """
        Set hash table value by key
        """
        bucket_path = self._get_bucket_path(key)
        bucket_lock = self._get_bucket_lock(bucket_path)
        with bucket_lock:
            bucket = json.load(bucket_path.open('r'))
            try:
                bucket[key] = val
                json.dump(bucket, bucket_path.open('w'))
            except Exception as e:
                raise e
        with self.log_lock:
            self._log(key, val)
